# Python

Antes de empezar queremos aclarar que esta intenta ser una guía relativamente rápida de inciación para Python. No vamos a explorar todas las funcionalidades que existen. Para complementar todo lo que presentamos a continuación, dejamos el link al tutorial oficial de Python 3 en español: https://docs.python.org/es/3/tutorial/ y a la referencia oficial también en español: https://docs.python.org/es/3/reference/index.html.


Vamos a trabajar con el formato de _notebooks_ en el cual se puede escribir pequeñas porciones (llamdas _celdas_) de código y ejecutarlas individualmente.

Un programa escrito en Python tiene su extensión caracteristica _.py_:

```programa_ejemplo.py```

Mientras que una notebook utiliza la extensión _.ipynb_:

```notebook_ejemplo.ipynb```




## Contenido de este Notebook

En este notebook trataremos los siguientes temas:

* Sintaxis Básica de Python
* Variables y Tipos de Datos
* Operadores artiméticos, de comparacióna y lógicos
* Control de Flujo con bloques if-else
* Métodos de Strings
* Importar Modulos Externos

## Ejectuando una celda
Abajo se encuentra una celda con la función _print()_, la cual se encarga de mostrar un texto en la salida.

Para ejecutar la celda solo hace falta hacer click en el pequeño triangulo que aparece a la izquierda de la celda cuando hacemos click dentro de esta.
Otra opción es usar el atajo _"Ctrl + Shift"._

In [ ]:
print("Hola Mundo")

Hola Mundo


Luego de ejecutar la celda, se podrá ver a la izquierda un ícono verde de una tilde si la ejecución se produjo sin problemas, y en caso contrario aparecerá un ícono de color rojo para indicarnos que hubo un fallo.

Justo debajo de la celda se mostrará el _output_ o salida de la celda ejectuada (el resultado de la ejecución).

Cada celda empieza con unos corchetes vacios [ ]. Luego de la primera ejecución aparecerá un número uno [1] para indicar que es la primera celda ejecutada, luego aparecerá un [2], un [3], etc, para recordarnos el orden de ejecución.

## Comentarios

Cuando se utiliza el símbolo "#", todo lo que siga a esa linea no será relevante para la ejecución de la celda, y es lo que se conoce como un _comentario_. Los comentarios pueden utilizarse para ir explicando el funcionamiento del programa. De esta forma, logramos dos cosas:  
1) nos manejamos mejor dentro del código del programa, sobre todo si volvemos a verlo después de un tiempo y no lo tenemos fresco  
2) facilitamos la colaboración con otras personas  
Por estas razones, está bueno escribir comentarios descriptivos en nuestros programas.

In [ ]:
# Esto es un comentario
print("Esto no es un comentario") # Con esta linea imprimimos el texto "Esto no es un comentario"

Esto no es un comentario


## Operaciones Matemáticas Simples

Una buena forma de empezar a usar python es probar hacer simples opraciones aritméticas. Es decir, vamos a usar pyhton como una calculadora sencilla. Mostramos algunos ejemplos y luego verán una tabla mostrando todos los operadores aritméticos en Python


In [ ]:
# suma
3 + 5

8

In [ ]:
# resta
8 - 12

-4

In [ ]:
# multiplicación
5 * 12

60

In [ ]:
# División
10 / 3

3.3333333333333335

*** Nota: si notan algo raro en el resultado de la división 10/3, no se preocupen que más adelante explicaremos la razón de este resultado

In [ ]:
# Exponente
5 ** 2 # así escribimos 5 al cuadrado ¡NO USAR 5^2! el operador ^ es válido pero no es el exponente

25

Aquí dejamos una lista completa de los operadores aritméticos

### Operadores Aritméticos

Símbolo | Operación  | Ejemplo
--------|------------|--------
`+` | Suma | `3 + 5` = 8
`-` | Resta | `8 - 5` = 3
`*` | Multiplicación | `3 * 4` = 12
`/` | División | `5 / 2` = 2.5
`**` | Exponente | `5**2` = 25
`//` | Parte entera de la división | `7 // 3` = 2
`%` | Resto de la división | `7 % 3` = 1

# Ejercicios

Si recién empiezan a programar o son nuevos en python, recomendamos ir haciendo los diferentes ejercicios propuestos a lo largo de estas notebooks.

Pueden usar la cenda donde se describe el ejercicio, o si les resulta más cómodo, pueden crear una nueva celda haciendo click en la opción "**+ Código**" arriba a la izquerda de la pantalla, o yendo a la opción en el menú: **Insertar -> Célda de código.**

In [ ]:
# EJERCICIO 1:

# Probar el resto de los operadores
# ¿Cuál cree que sería el resultado de 10 // 3?
# ¿Y Cual sería el resultado de 10 % 3?
# ¿Puede entender la diferencia entre estos operadores y el operador de división?


In [ ]:
# EJERCICIO 2:

# Pruebe ejecutar la expresión 5^2 para verificar que no es 5 al cuadrado.
# ¿Cómo puedo obtener el resultado de 5 al cuadrado?

## Expresiones más complejas

Todo lo que fuimos escribiendo en las celdas anteriores son expresiones. En python una expresión es una combinación de valores, operadores, y llamadas a funciones que se evalúan a un cierto valor. Por ejemplo si se evalúa la expresión `2 + 2` se obtiene el valor 4.

Las expresiones que vimos anteriormente son bastante simples, pero pueden empezar a complicarse.

Por ejemplo, es claro que `2 + 2` es 4, o que `9 / 3` es 3, pero no es tan fácil saber a simple vista a que valor se evalua la expresion `5 + 5*10/2**2`.

Primero debemos tener en cuenta que el orden de los operadores es el mismo que aprendimos en matemáticas el secundario, y segundo, que podemos usar paréntesis para cambiar este orden, o simplemente para organizar de forma más clara una expresión.

### Orden de los operadores

El orden de las operaciones el el siguiente

1. Exponente (**)
2. Multiplicación (*), División (/), Parte entera de la división (//), Resto (modulo) (%)
3. Suma (+), Resta (-)

In [ ]:
# Suma y Producto. Primero se resuelve el producto y después la suma
5 + 3*4

17

In [ ]:
# Suma y producto. Usando paréntesis podemos hacer que se resuelva primero la suma y luego el producto
(5+3) * 4

32

## Errores ejectuando una celda

**No siempre se ejecturará correctamente una celda**. Por ejemplo ¿Qué pasa si intento dividir un número por cero?


In [ ]:
# EJERCICIO

# Dividir un número por cero y ver que pasa.
# Por ejemplo: 5 / 0

Intentar dividir un número sobre cero nos devuelve un error. Especificamente el error llamado **ZeroDivisionError**. Al final de la salída de la celda podemos ver el nombre del error junto a una breve descripción del mismo:

`ZeroDivisionError: division by zero`

Existen multiples tipos de errores con los cuales nos iremos familiarizando a medida que sigamos programando.

## Variables

Supongamos que queremos calcular el módulo de la fuerza entre dos particulas cargadas $q_1$ y $q_2$, separadas una distancia $r$ entre sí. Para esto utilizamos la ley de coulomb (despreciando la fuerza gravitacional). Es decir:

$$ F = k \frac{q_1 \cdot q_2}{r^2}$$

Si tenemos los siguientes datos con unidades en el SI:

$$ k = 8.9\times 10^{9}$$
$$ q_1 = 0.9 \times 10^{-6}$$
$$q_2 = 1.2 \times 10^{-6}$$
$$r = 2.5 \times 10^{-3}$$

podemos calcular la fuerza escribiendo la siguiente expresión:



In [ ]:
# Escribimos la expresión para calcular la fuerza electrostáctica
8.9 * 10**9 * (0.9*10**-6 * 1.2*10**-6) / ((2.5 * 10**-3)**2)

1537.9199999999998

O mejor aún, podemos usar variables para cada uno de nuestros valores, de la siguiente forma:

In [ ]:
# Primero creamos variables para cada uno de nuestros datos
# Notemos que en vez de escribir 8.9 * 10**9 escribimos 8.9E9.
# Esta notación es equivalente y más sencilla de usar
k = 8.9e9
q_1 = 0.9e-6
q_2 = 1.2e-6
r = 2.5e-3

# Ahora podemos calcular la fuerza de una forma mucho más sencilla
# y con menos posibilidades de error
F = k * (q_1*q_2)/(r**2) # También podemos guardar la fuerza en otra variable
print(F)

1537.9199999999998


En python el símbolo = denota el operador de asignación (no representa igualdad necesariamente, sino asignación, ya hablaremos de eso más adelante).

De forma general creamos una variable de la siguiente forma:

`nombre_de_mi_variable = expresión`

Es decir, elegimos un nombre para nuestra variable, usamos el operador = y luego una expresión válida.

Algunas reglas importantes al momento de elegir el nombre de una variable son que no podemos empezar el nombre con un número y que no podemos dejar espacios en medio.

Luego, existen algunas convenciones que debemos conocer.
Por ejemplo, debido a que no se pueden usar espacios en el nombre de una variable, existen diferentes formas de denotar la separación entre palabras. Dos convenciones muy utilizadas son:

* Camel Case: estaEsMiVariable
* Snake Case: esta_es_mi_variable

En python la convención es usar snake case, reemplazando los espacios por un guión bajo. También se desaconseja usar acentos o la letra ñ.


In [ ]:
# volvemos a nuestro primer ejemplo, pero esta vez usando una variable
mensaje = "hola mundo"
print(mensaje)

hola mundo


Las variables nos sirven para guardar valores que podamos necesitar más adelante, y para mejorar la comprensión de nuestro código.

Considerando nuestro ejemplo anterior de la fuerza electromagnética, si vemos esta línea de código es dificil tener claro qué estamos haciendo:

`8.9 * 10**9 * (0.9*10**-6 * 1.2*10**-6) / ((2.5 * 10**-3)**2)`

Mientras que de esta forma tiene mucho más sentido:

`F = k * (q_1*q_2)/(r**2)`

***NOTA:*** Siempre que podamos es imporante dar un nombre descriptivo a nuestras variables. Normalmente de desaconseja usar letras únicas para declarar variables. En el caso de problemas de física podemos hacer una excepción ya que muchas constantes son expresadas de esta forma, pero debemos tener en cuenta que el nombre de la variable debe ser descriptivo de alguna forma.

In [ ]:
# podemos reasignar el valor de una variable en cualquier momento del código.
mi_nivel_de_estres = 1
print(mi_nivel_de_estres)

# Reasignamos la variable con otro valor
mi_nivel_de_estres = 5
print(mi_nivel_de_estres)

1
5


In [ ]:
# Nota: Podemos pasar varios argumentos a la función print:
# por ejemplo:
print("Mi nivel de estres es:", mi_nivel_de_estres)

Mi nivel de estres es: 3


In [ ]:
# También podemos actualizar nuestra variable dependiendo del valor que tenía anteriormente
mi_nivel_de_estres = 1
print("Mi nivel de estres es:", mi_nivel_de_estres)

# Recordemos que el signo = no es el igual matemático, sino un operador de asignación
# Sí, podemos hacer esto:
mi_nivel_de_estres = mi_nivel_de_estres + 1
print("Mi nivel de estres es:", mi_nivel_de_estres)
# Cuando asignamos un valor a una variable, siempre se resuelve primero la expresión
# a la derecha del igual, y luego recién se realiza la asignación.

# Otra forma de hacer lo mismo es:
mi_nivel_de_estres +=  1
print("Mi nivel de estres es:", mi_nivel_de_estres)

Mi nivel de estres es: 1
Mi nivel de estres es: 2
Mi nivel de estres es: 3


## Tipos de Datos

Hasta ahora estuvimos trabajando principalmente con números, pero existen otros tipos de datos.
Para saber con que tipo de dato estamos trabajando podemos usar la función type().

### Datos Numéricos
Debemos distinguir entre los valores enteros (int) y los valores de punto flotante (float):

In [ ]:
a = 1   # int
b = 1.1 # float

# para saber el tipo de una variable podemos usar una función precargada llamada type():
print(type(a))
print(type(b))

<class 'int'>
<class 'float'>


In [ ]:
# si tenemos una operación entre un entero y un flotante, el resultado es
# otro flotante, ya que así no perdemos información
var = 5 + 5.5
type(var)

float

In [ ]:
# En una división de enteros, el resultado también es un flotante
type(5/2)

float

También existen los números complejos como la suma de la parte real y la parte imaginaria (con la letra j denotando la parte imaginaria)

In [ ]:
es_complejo = 1+2j
# podemos acceder a la parte real e imaginaria individualmente con la notación
# de punto, de la siguiente forma:
print("parte real:", es_complejo.real)
print("parte imaginaria:", es_complejo.imag)
type(es_complejo)

parte real: 1.0
parte imaginaria: 2.0


complex

### Cadenas de caracteres

Los caracteres que están entre comillas se llaman Cadenas o *strings* y se usan generalmente para representar texto.


Pära crear una cadena o string utilizamos comillas dobes " o comillas simples '.

`cadena_valida = "mi texto aquí"`

`otra_cadena_valida = 'más texto aquí'`

Si abrimos con comillas simples, debemos cerrar con comillas simples, y si abrimos con comillas dobles, debemos cerrar con comillas dobles. Es decir que esto no es válido:

`cadena_invalida = "un texto'`


In [ ]:
FACET = "Facultad de Ciencias Exactas y Tecnología"
type(FACET)

str

In [ ]:
esto_no_es_una_suma = "1 + 5" # esto deja de ser una operación numérica al colocar las comillas
type(esto_no_es_una_suma)

str

In [ ]:
# Si intentamos sumar dos cadenas, se concatenan una detrás de la otra:
suma_de_cadenas = "hola " + "como están?"
print(suma_de_cadenas)

hola como están?


Usar operaciones aritméticas mezclando cadenas y numéros generalmente va a generar un error, con la exepción de la multiplicación entre un entero y una cadena.

In [ ]:
# así nos podemos reir fuerte:
print(10*"JA")

# y así más despacio:
print(3*"ja")

JAJAJAJAJAJAJAJAJAJA
jajaja


A veces es necesario utilizar tanto las comillas simples como las dobles. Por ejemplo si dentro del texto queremos que haya comillas dobles:

In [ ]:
print('Estoy usando "comillas" dobles dentro de esta string')

Estoy usando "comillas" dobles dentro de esta string


In [ ]:
# EJERCICIO:

# Como puedo imprimir en una sola linea el siguiente mensaje
# que nos muiestra dos tipos de comillas:
# comillas doble: "", comillas simple: ''

### Booleanos

Otro tipo de dato son los booleanos. En este caso solo hay dos valores posibles:

In [ ]:
True  # True (con T mayúscula)

True

In [ ]:
False # False (con F mayúscula)

False

In [ ]:
# Podemos usar True y False dentro de operaciones aritméticas.
# True toma el valor de 1 y False toma el valor 0:

print(5 + True)
print(False*7)

6
0


Más adelante veremos como se utilizan los Booleanos para el control de flujo

### Valor Nulo

Por último también existe el dato nulo. Este tipo de dato denota la ausencia de un valor, puede parecer poco útil, pero ya veremos algunos ejemplos con él

In [ ]:
aqui_no_hay_valor = None

type(aqui_no_hay_valor)

NoneType

### Convertir un tipo de dato a otro (Casting)

Existen funciones que nos permiten pasar de un tipo de dato a otro

float(), str(), int(), bool()

In [ ]:
# al transformar un flotante a un entero se pierde la parte decimal, NO se redondea!
int(5.9)

5

In [ ]:
float(8)

8.0

In [ ]:
str(42)

'42'

In [ ]:
# Todos los números distintos a 0 se convierten en True, cero se convierte en False
bool(-1)

True

In [ ]:
# Las cadenas también se evaluan a True, con exepción de la cadena vacia ""
bool("")

False

In [ ]:
bool("cadena")

True

In [ ]:
# podemos convertir una string a un valor numérico en caso de que la string
# contenga solamente números:
string_1 = "5"
string_2 = "4"
print("Aquí estamos concatenando dos strings:", string_1 + string_2)
print("Aquí estamos sumando dos enteros:", int(string_1) + int(string_2))

Aquí estamos concatenando dos strings: 54
Aquí estamos sumando dos enteros: 9


## Un poco de control de flujo

Muchas veces queremos que nuestro programa realice diferentes acciones dependiendo de ciertos factores.

Empezaremos aprendiendo sobre los operadores de comparación:


### Operadores de Comparación

Los operadores de comparación nos resultarán bastante familiares. Una comparación se realiza entre dos objetos, y si la expresión es válida el resultado será un booleano (True o False)

In [ ]:
5 < 4 # False

False

In [ ]:
12 > 5 # True

True

In [ ]:
# También existen el mayor o igual >= y menor o igual <=
5 >= 4

True

In [ ]:
# El doble == es el operador de igualdad
5 == 4

False

In [ ]:
# Mientras que != implica desigualdad
12 != 12.5

True

#### Problemas de representar números reales

Veamos el siguiente ejemplo

In [ ]:
# La siguiente expresión se evalua a False, aunque debería ser True
0.3 == 0.4 - 0.1

False

Antes vimos que el resultado obtenido de `10/3` era `3.3333333333333335`.

El problema es el 5 que aparece al final. Claramente no debería estar ahi.

Para entender esto necesitamos tener en cuenta que los números reales son infinitos, pero la computadora no tiene memoria infinita, por lo tanto no puede representar números reales, sino que los aproxima.

Solo ciertos números pueden expresarse de forma exacta con los números flotantes.

Estos números son aquellos que puedan expresarse como una suma de fracciones donde los denominadores son múltiplos de dos, de la siguiente forma:

$$\frac{a}{2^0} + \frac{b}{2^1} + \frac{c}{2^2} + \cdots$$

Por ejemplo el número 1.75 se puede expresar como:

$$
 1.75 =  \frac{1}{2^0} + \frac{3}{2^2}
$$

Sin embargo 10/3 no puede representarse de manera exacta de esta forma, por lo tanto el resultado es una aproximación

In [ ]:
# Estos números pueden representarse de forma exacta, por lo tanto no hay ningún error
1.75 == 1 + 3/4

True

En general se recomienda NO usar el operador `==` para comparar dos flotantes

### Sentencia if - else

La sentencia if nos permite ejecutar un bloque de código dependiendo de una condición.

```
if <expresión>:
    bloque 1
else:
    bloque 2
```

Si la expresión se evalúa a True, se ejecuta el bloque 1, de lo contrario se ejecuta el bloque 2.
Es importante notar la indentación (el esapacio o la sangría) en los bloques 1 y 2.
Esta indentación es necesaria para la correcta ejecución de los bloques de código.

In [ ]:
# Un ejemplo:
edad = 21

if edad >= 18:
    print("¡Puedo tomar alcohol legalmente!")
else:
    print("No puedo tomar alcohol legalmente :(")

# este código esta fuera de ambos bloques, por lo tanto siempre se ejecutará
print("Tomen con moderación gente")


¡Puedo tomar alcohol legalmente!
Tomen con moderación gente




#### Función input():

Para hacer más interesante nuestro ejemplo anterior podemos usar la función `input()` para crear una variable que será definida en tiempo de ejecución por el usario.

Primero veamos un ejemplo:

In [ ]:
print("Cuál es tu nombre? ")
nombre = input()

print("Hola", nombre)

Cuál es tu nombre? 
Atuel
Hola Atuel


In [ ]:
# Ahora, queremos que nuestro programa nos pregunte que edad tenemos, pero hay un problema
edad = input("Que edad tenés? ',:c")
type(edad)

str

la función `input()` siempre devuelve una string. Por lo tanto para hacer la comparación necesitamos transformar la string en un entero:

In [ ]:
edad = int(input("Que edad tenés? ',:c"))

if edad >= 18:
    # Este es el bloque del if
    print("¡Puedo tomar alcohol legalmente!")
else:
    # Este es el bloque del else
    print("No puedo tomar alcohol legalmente :(")

# este código esta fuera de ambos bloques, por lo tanto siempre se ejecutará
print("Tomen con moderación gente")


No puedo tomar alcohol legalmente :(
Tomen con moderación gente


### Operadores Lógicos

Otro tipo de operadores bastante útiles en estos casos son los operadores lógicos:

Tenemos tres operadores lógicos: `and`, `or` y `not`

Operación | Resultado
--------|------------
`a and b`| Si `a` se evalua `False` devuelve el valor de `a`, sino el de `b`
`a or b` | Si `a` se evalua `False` devuelve el valor de b, sino el de `a`
`not a` | Si `a` se evalua `False` devulve `True`, si no devuelve `False`

Estas explicaciones pueden parecer confusas, sobre todo para los operadores `and` y `or`, pero tengamos en cuenta lo siguiente;
 si `a` y `b` son booleanos (`True` o `False`) entonces se cumple que:

Operación | Resultado
--------|------------
`a and b`| Se evalua a `True` solo si tanto a como b son `True`, sino `False`
`a or b` | Se evalua a `True` si a es` True`, o si b es `True`, o si ambos son `True`, de lo contrario False

pero veamoslo con un ejemplo:

In [ ]:
# Tenemos dos variables, la edad y una variable que indica si estamos o no manejando
edad = 19
manejando = True

# Solo vamos a poder tomar si se cumplen dos condiciones (tener más de 18 y no estar manejando)
if edad > 18 and not manejando:
    print("Puede tomar")
else:
    print("No puede tomar")


No puede tomar


#### **Ejercicio Opcional**: entendiendo el funcionamiento de los operadores `and` y `or`

Veamos que pasa cuando usamos `and` y `or` con valores no booleanos.


In [ ]:
# EJERCICIO

# ¿Cual es el resultado de las siguientes expresiones?
# 1 and 2
# 0 and 1
# 1 or 2
# 0 or 2
# "a" or "b"
# "" and "b"

¿Puede entender porque estos operadores funcionan de esta forma?

In [ ]:
max(1, 2, 4)  # devuelve el valor máximo

4

In [ ]:
min(2, -5, -10) # devuelve el valor mínimo

-10

In [ ]:
abs(-55)  # devuelve el valor absoluto

55

## Strings con formato

Las strings con formato nos permiten colocar variables dentro de una string.

Supongamos que queremos imprimir un mensaje del tipo:

"Hola Juan!"

pero no queremos saludar siempre a Juan, sino a una persona genérica.

In [ ]:
# Una opción es realizar una concatenación de strings:
nombre = "Blas"
print("Hola " + nombre + "!")

# Pero la alternativa más cómoda es utilizar f-strings:
nombre = "Blas"
print(f"Hola {nombre}!")

Hola Blas !
Hola Blas!


las f-strings empiezan con una letra **f** situada justo antes de las comillas. Luego las variables se colocan entre dos llaves {}

In [ ]:
# otro ejemplo
numero = 5
objeto = "botellas"
lugar = "patio"

mensaje = f"hay {numero} {objeto} en el {lugar}" # probar sacar la f delante de la string y ejecutar
print(mensaje)

hay 5 botellas en el patio


Dentro de una f-string también podemos dar formato a los números que vayamos a imprimir en pantalla

In [ ]:
num = 31.311

print(f"{num:.3f}") # número con formato flotante, con tres cifras luego de la coma
print(f"{num:.1f}") # una cifra luego de la coma
print(f"{num:.2e}") # notación exponencial, con 2 cifras luego e la coma

31.311
31.3
3.13e+01


## Métodos de Strings

Los métodos son funciones que pertenecen a un objeto. Para acceder a un método tenemos que partir desde el objeto usando la notación de punto:

`string.metodo()`

Algunos métodos utiles para las strings son:

* `upper()` : Devuelve la string en mayúsculas
* `lower()` : Devuelve la string en minúsculas
* `strip()` : Devuelve la string sin espacios al final ni al inicio
* `title()` : Devuelve una string con cada primer letra en mayúscucla
* `replace()` : Permite reemplazar un valor con otro

Veamos algunos ejemplos:

In [ ]:
"ESTOY EN MAYUSCULAS!".lower()

'estoy en mayusculas!'

In [ ]:
cadena = "vamos de viaje!"

# cambiamos viaje por fiesta:
cadena.replace("viaje", "fiesta")

'vamos de fiesta!'

## Otras funciones:
Existen varias funciones precargadas en python puro, a continuación mostramos algunas que pueden resultarnos útiles:

In [ ]:
# devuelve el valor máximo:
max(1, 2)

2

In [ ]:
# devuelve el valor mínimo
min(1, 5, -10)

-10

In [ ]:
# valor absoluto
abs(-75)

75

In [ ]:
# redondear
round(5.88)

6

In [ ]:
# redondear con una cifra decimal
round(5.88, 1)

5.9

#### Comparar flotantes con `round()`

Usando la función round() podemos comparar la igauldad dos flotantes de forma más segura

In [ ]:
0.3 == round(0.4 - 0.1, 1)

True

## Importar Librerias

Al instalar python también tenemos acceso a su biblioteca estandar. Dentro de esta biblioteca tenemos varios módulos y librerias (conjuntos de módulos) que vienen con sus propias funciones, tipos de datos y variables. A continuación mostramos dos modulos bastante útiles; `math` y `random`:

Para utilizar un módulo usamos la palabra `import` seguida del nombre del módulo:

```
import <nombre-del-modulo>
```

### Módulo `math`

El modulo math tiene cargadas muchas funciones matemáticas y algunas variables como pi.

### Módulo `random`

El módulo random nos permite generar números ["pseudoaleatorios" ](https://es.wikipedia.org/wiki/N%C3%BAmero_pseudoaleatorio#:~:text=Un%20n%C3%BAmero%20pseudoaleatorio%20es%20un,pero%20no%20lo%20hace%20realmente.). Estos son números que a fines practicos simular ser números aleatorios.

### Estilo de Importación

Hay diferentes formas de importar y usar un modulo


In [ ]:
# Opción 1. Usar la notación de punto
# Esta es la opción más sencilla. Luego de importar debemos usar un punto seguido del objeto que vayamos a utilizar
import math

# seno de pi/2:
math.sin(math.pi/2)

1.0

In [ ]:
# Opción 2. Importar solo los elementos que vayamos a usar
from math import sin, pi

# seno de pi/2:
sin(pi/2)

1.0

In [ ]:
# Opción 3. No muy recomendada, pero válida para casos particulares
from math import *

# seno de pi/2:
sin(pi/2)
# El problema de este método es que puede haber un conflicto entre dos o más módulos
# Por lo tanto en caso de estar usando varios modulos o librerias no se recomienda usar
# esta notación, ya que puede haber funciones con el mismo nombre en ambos lugares

1.0

In [ ]:
# Opción 4. Usar un alias
# Igual a la opçión 1 pero usando un alias o apodo para el nombre del modulo
import math as m

# seno de pi/2:
m.sin(m.pi/2)

1.0

Algunas funciones útiles en math son:

`cos(x)` : coseno de x

`sin(x)` : seno de x

`log(x)` : logaritmo natural de x

`log10(x)` : logaritmo base 10 de x

`sqrt(x)` : raiz cuadrada de x

`exp(x)` : e elevado a la x

ver https://docs.python.org/es/3/library/math.html para más información.

In [ ]:
import random

# la función random() devuelve un valor flotante aleatorio del intervalo semiabierto [0, 1)
random.random()

0.9659202588640557

In [ ]:
a = 1
b = 3

# la función randint(a, b) devuelve un valor entre a y b (incluidos a y b)
random.randint(a, b)

1

Mas adelante veremos más funciones del módulo `random`